<h1 id="deep-mutational-scanning-of-sars-cov-2-kp311-spike-using-a-barcoded-lentiviral-platform">Deep mutational scanning of SARS-CoV-2 KP.3.1.1 spike using a barcoded lentiviral platform</h1>
<p>Analysis by Bernadeta Dadonaite and Jesse Bloom (2024)</p>
<p>See <a href="https://github.com/dms-vep/SARS-CoV-2_KP.3.1.1_spike_DMS">https://github.com/dms-vep/SARS-CoV-2_KP.3.1.1_spike_DMS</a> for full code.</p>
<div class="toc"><span class="toctitle">Contents</span><ul>
<li><a href="#barcode-to-codon-variant-lookup-table">Barcode to codon-variant lookup table</a></li>
<li><a href="#count-barcodes-for-variants">Count barcodes for variants</a></li>
<li><a href="#functional-effects-of-mutations">Functional effects of mutations</a></li>
</ul>
</div>
<h2 id="barcode-to-codon-variant-lookup-table">Barcode to codon-variant lookup table</h2>
<h4 id="analysis-notebooks">Analysis notebooks</h4>
<ul>
<li><a href="notebooks/analyze_pacbio_ccs.html">Analysis of PacBio CCSs</a></li>
<li><a href="notebooks/build_pacbio_consensus.html">Building barcoded variant consensus sequences</a></li>
<li><a href="notebooks/build_codon_variants.html">Building of codon-variant table</a></li>
</ul>
<h4 id="data-files">Data files</h4>
<ul>
<li><a href="https://github.com/dms-vep/SARS-CoV-2_KP.3.1.1_spike_DMS/blob/main/results/gene_sequence/codon.fasta">parental codon sequence</a></li>
<li><a href="https://github.com/dms-vep/SARS-CoV-2_KP.3.1.1_spike_DMS/blob/main/results/gene_sequence/protein.fasta">parental protein sequence</a></li>
<li><a href="https://github.com/dms-vep/SARS-CoV-2_KP.3.1.1_spike_DMS/blob/main/results/variants/codon_variants.csv">barcode to codon-variant lookup table</a></li>
</ul>
<h2 id="count-barcodes-for-variants">Count barcodes for variants</h2>
<h4 id="analysis-notebooks_1">Analysis notebooks</h4>
<ul>
<li><a href="notebooks/analyze_variant_counts.html">Analysis of variant counts</a></li>
</ul>
<h4 id="data-files_1">Data files</h4>
<ul>
<li><details><summary>Barcode count CSVs (click triangle to expand/collapse)</summary><ul>

<li><a href="https://github.com/dms-vep/SARS-CoV-2_KP.3.1.1_spike_DMS/blob/main/results/barcode_counts/Lib1-250213-VSVG_control_293T_counts.csv">Lib1-250213-VSVG_control_293T</a></li>
<li><a href="https://github.com/dms-vep/SARS-CoV-2_KP.3.1.1_spike_DMS/blob/main/results/barcode_counts/Lib2-250213-VSVG_control_293T_counts.csv">Lib2-250213-VSVG_control_293T</a></li>
<li><a href="https://github.com/dms-vep/SARS-CoV-2_KP.3.1.1_spike_DMS/blob/main/results/barcode_counts/Lib1-250213-no-antibody_control_293TACE2_counts.csv">Lib1-250213-no-antibody_control_293TACE2</a></li>
<li><a href="https://github.com/dms-vep/SARS-CoV-2_KP.3.1.1_spike_DMS/blob/main/results/barcode_counts/Lib2-250213-no-antibody_control_293TACE2_counts.csv">Lib2-250213-no-antibody_control_293TACE2</a></li>
</ul></details>

</li>
</ul>
<h2 id="functional-effects-of-mutations">Functional effects of mutations</h2>
<h4 id="final-summary-plots">Final summary plots</h4>
<ul>
<li><details><summary>Interactive plots of average mutation functional effects (click triangle to expand/collapse)</summary><ul>

<li><a href="htmls/293T_high_ACE2_entry_func_effects.html">293T_high_ACE2_entry</a></li>
</ul></details>

</li>
<li><details><summary>Interactive plots of average mutation latent-phenotype effects (click triangle to expand/collapse)</summary><ul>

<li><a href="htmls/293T_high_ACE2_entry_latent_effects.html">293T_high_ACE2_entry</a></li>
</ul></details>

</li>
</ul>
<h4 id="analysis-notebooks_2">Analysis notebooks</h4>
<ul>
<li><a href="notebooks/analyze_func_scores.html">Analysis of functional scores</a></li>
<li><details><summary>Notebooks with per-selection global epistasis fitting (click triangle to expand/collapse)</summary><ul>

<li><a href="notebooks/func_effects_global_epistasis_Lib1-250213_high_ACE2.html">Lib1-250213_high_ACE2</a></li>
<li><a href="notebooks/func_effects_global_epistasis_Lib2-250213_high_ACE2.html">Lib2-250213_high_ACE2</a></li>
</ul></details>

</li>
<li><details><summary>Notebooks averaging mutation functional effects (click triangle to expand/collapse)</summary><ul>

<li><a href="notebooks/avg_func_effects_293T_high_ACE2_entry.html">293T_high_ACE2_entry</a></li>
</ul></details>

</li>
</ul>
<h4 id="data-files_2">Data files</h4>
<ul>
<li><details><summary>Per-variant functional score CSVs (click triangle to expand/collapse)</summary><ul>

<li><a href="https://github.com/dms-vep/SARS-CoV-2_KP.3.1.1_spike_DMS/blob/main/results/func_scores/Lib1-250213_high_ACE2_func_scores.csv">Lib1-250213_high_ACE2</a></li>
<li><a href="https://github.com/dms-vep/SARS-CoV-2_KP.3.1.1_spike_DMS/blob/main/results/func_scores/Lib2-250213_high_ACE2_func_scores.csv">Lib2-250213_high_ACE2</a></li>
</ul></details>

</li>
<li><details><summary>Per-selection mutation functional effect CSVs (click triangle to expand/collapse)</summary><ul>

<li><a href="https://github.com/dms-vep/SARS-CoV-2_KP.3.1.1_spike_DMS/blob/main/results/func_effects/by_selection/Lib1-250213_high_ACE2_func_effects.csv">Lib1-250213_high_ACE2</a></li>
<li><a href="https://github.com/dms-vep/SARS-CoV-2_KP.3.1.1_spike_DMS/blob/main/results/func_effects/by_selection/Lib2-250213_high_ACE2_func_effects.csv">Lib2-250213_high_ACE2</a></li>
</ul></details>

</li>
<li><details><summary>Average mutation functional effects (CSV files) (click triangle to expand/collapse)</summary><ul>

<li><a href="https://github.com/dms-vep/SARS-CoV-2_KP.3.1.1_spike_DMS/blob/main/results/func_effects/averages/293T_high_ACE2_entry_func_effects.csv">293T_high_ACE2_entry</a></li>
</ul></details>

</li>
<li><details><summary>Average mutation latent-phenotype effects (CSV files) (click triangle to expand/collapse)</summary><ul>

<li><a href="https://github.com/dms-vep/SARS-CoV-2_KP.3.1.1_spike_DMS/blob/main/results/func_effects/averages/293T_high_ACE2_entry_func_effects.csv">293T_high_ACE2_entry</a></li>
</ul></details>

</li>
</ul>